proposals/infra/INF-0002-root-signature-in-clang.md

```diff
@@ -0,0 +1,697 @@
+  hlsl::RootSignature::TextureAddressMode AddressV =
+      hlsl::RootSignature::TextureAddressMode::Wrap;
+  hlsl::RootSignature::TextureAddressMode AddressW =
+      hlsl::RootSignature::TextureAddressMode::Wrap;
+  float MipLODBias = 0;
+  uint32_t MaxAnisotropy = 16;
+  hlsl::RootSignature::ComparisonFunc ComparisonFunc =
+      hlsl::RootSignature::ComparisonFunc::LessEqual;
+  hlsl::RootSignature::StaticBorderColor BorderColor =
+      hlsl::RootSignature::StaticBorderColor::OpaqueWhite;
+  float MinLOD = 0;
+  float MaxLOD = hlsl::RootSignature::Float32Max;
+  uint32_t ShaderRegister;
+  uint32_t RegisterSpace = 0;
+  hlsl::RootSignature::ShaderVisibility ShaderVisibility =
+      hlsl::RootSignature::ShaderVisibility::All;
+};
+
+struct ParsedRootSignature {
+  dxbc::RootSignature::ContainerRootSignatureDesc RSDesc;
+  llvm::SmallVector<RootSignature::ContainerRootParameter, 8>
+      RSParameters;
+  llvm::SmallVector<RootSignature::StaticSamplerDesc, 8> StaticSamplers;
+  llvm::SmallVector<RootSignature::ContainerDescriptorRange, 8>
+      DescriptorRanges;
+  llvm::SmallVector<
+      std::variant<RootSignature::ContainerRootConstants,
+                   RootSignature::ContainerRootDescriptor,
+                   RootSignature::ContainerRootDescriptorTable>,
+      8>
+      RSParamExtras;
+};
+
+}
+
+```
+
+This will require more work for print and serialization the AST node.
+Since DirectX backend also need to print and serialzation root signature, 
+the code should be able to share between the frontend and backend.
+The code to serialize a HLSLRootSignatureDecl will be looked like this:
+
+```
+
+void ASTDeclWriter::VisitHLSLRootSignatureDecl(HLSLRootSignatureDecl *D) {
+  VisitNamedDecl(D);
+  Record.AddSourceLocation(D->getKeywordLoc());
+  SmallVector<uint32_t, 64> Data = hlsl::serializationRootSignature(D->getRootSignature());
+  Record.writeUInt32(Data.size());
+  for (uint32_t Val : Data)
+    Record.push_back(Val);
+  Code = serialization::DECL_HLSL_ROOT_SIGNATURE;
+}
+
+
+```
+
+If we don't want to share the serializationRootSignature code.
+The AST node serialization code might be looked like this:
+```
+void ASTDeclWriter::VisitHLSLRootSignatureDecl(HLSLRootSignatureDecl *D) {
+  VisitNamedDecl(D);
+  Record.AddSourceLocation(D->getKeywordLoc());
+  hlsl::ParsedRootSignature &RS = D->getRootSignature();
+  Record.push_back(RS.RSDesc.Version);
+  Record.push_back(RS.RSDesc.Flags);
+  for (const auto &P : ParsedRS.RSParameters) {
+    // push_back all the fields of root parameters.
+  }
+  for (const auto &SS : ParsedRS.StaticSamplers) {
+    // push_back all the fields of static samplers.
+  }
+  for (const auto &Range : ParsedRS.DescriptorRanges) {
+    // push_back all the descriptor ranges.
+  }
+  
+  Code = serialization::DECL_HLSL_ROOT_SIGNATURE;
+}
+
+```
+Unless there's other correct way to serialize the data, share serialization 
+code between clang and llvm seems better.
```

Comment:
The `clang::ASTRecordWriter` writes to `clang::ASTWriter::RecordDataImpl` which is a `llvm::SmallVectorImpl<uint64_t>`. The backing `llvm::SmallVector<uint64_t>` then gets encoded with the bitcode writer.

If you serialize using the backend serialization you're then going to need to convert it to an array of uint64_t's and let the bitcode serializer operate on those. That doesn't make sense from an architecture perspective and it doesn't seem to save any complexity to me.

Reply:
Updated to use Record.AddString.

Reply:
So the correct way is pack uint64_t for all the uint32_t data in a ParsedRootSignature?

Could you add pseudo code in your mind to help understanding?

Reply:
The code you have above here does what I'm describing. When you call `push_back` to append each value into the ASTWriter you're adding them to the vector that gets serialized in the bitcode. When the bitcode writer writes the series of values it bitpacks the values into less storage as I described above.

That is the correct way to serialize binary data that is part of an AST node, rather than any method that could share the serialization code with LLVM.

Other than the comment in the spec here you've provided no explanation of how you would share serialization with LLVM.

Reply:
I see.
If we share serialization code with LLVM, extra work will be done for build the offset data and store those offsets.  If not share, those could be saved for AST serialization.

Reply:
I don’t understand. We shouldn’t encode the offsets into the AST serialization, they would be meaningless since the data gets bit-packed by the bitcode writer.

Reply:
Add the issue for share code to the doc.

Reply:
Are you still considering sharing the serialization with the DirectX backend as a viable option? If so can you please include information in the spec about how you would do that. What actually gets encoded in the serialized AST?

Reply:
No. I'll remove that.